packages/builder/builder/src/plugins/sourceBuild.ts

```diff
@@ -123,22 +124,80 @@ export function builderPluginSourceBuild(
       const projectRootPath = api.context.rootPath;
 
       // TODO: when rspack support tsconfig paths functionality, this comment will remove
-      //   if (api.context.bundlerType === 'rspack') {
-      //     (api as RspackBuilderPluginAPI).modifyRspackConfig(async config => {
-      // // when support chain.resolve.conditionNames API, remove this logic
-      //       setConfig(config, 'resolve.conditionNames', [
-      //         '...', // Special syntax: retain the original value
-      //         sourceField,
-      //         ...(config.resolve?.conditionNames ?? []),
-      //       ]);
-      //     });
-      //   }
+      // if (api.context.bundlerType === 'rspack') {
+      //   (api as RspackBuilderPluginAPI).modifyRspackConfig(async config => {
+      //     // when support chain.resolve.conditionNames API, remove this logic
+      //     setConfig(config, 'resolve.conditionNames', [
+      //       '...', // Special syntax: retain the original value
+      //       sourceField,
+      //       ...(config.resolve?.conditionNames ?? []),
+      //     ]);
+      //   });
+      // }
       if (api.context.bundlerType === 'webpack') {
-        sourceBuildInWebpack(api as WebpackBuilderPluginAPI, {
-          projectRootPath,
-          sourceField,
-          projectName,
-          extraMonorepoStrategies,
+        (api as WebpackBuilderPluginAPI).modifyBuilderConfig(async config => {
+          const { sourceBuild = true } = config.experiments ?? {};
+
+          if (!sourceBuild) {
+            return;
+          }
+          const includes = await getSourceInclude({
+            projectNameOrRootPath: projectName || projectRootPath,
+            sourceField,
+            findMonorepoStartPath: projectRootPath,
+            extraMonorepoStrategies,
+          });
+          config.source = config.source ?? {};
+          config.source.include = [
+            ...(config.source.include ?? []),
+            ...includes,
+          ];
+        });
+
+        api.modifyBundlerChain((chain, { CHAIN_ID }) => {
+          const {
+            experiments: { sourceBuild },
+          } = (api as WebpackBuilderPluginAPI).getNormalizedConfig();
+
+          if (!sourceBuild) {
+            return;
+          }
+          // webpack.js.org/configuration/module/#ruleresolve
+          chain.module
+            .rule(CHAIN_ID.RULE.JS)
+            .resolve.mainFields.merge(['...', sourceField]);
+        });
+
+        (api as WebpackBuilderPluginAPI).modifyWebpackConfig(config => {
+          const {
+            experiments: { sourceBuild },
+          } = (api as WebpackBuilderPluginAPI).getNormalizedConfig();
+
+          if (!sourceBuild) {
+            return;
+          }
+
+          const rules = config.module?.rules ?? [];
+          const jsRule = rules.filter(rule => {
+            if (
```

Comment:
If webpack-chain does not support setting the conditionNames field, then we can try using the `.merge` method of webpack chain. 

Example: https://github.com/web-infra-dev/modern.js/blob/main/packages/builder/builder-webpack-provider/src/plugins/basic.ts#L25

It is not recommended to use config to traverse because it is very unstable. .